### PR TITLE
fix(ci): add pre_download_compilers

### PR DESCRIPTION
**Pull request by tomg10**

### CODE REVIEW

**Review by hatemosphere — DISMISSED** [review]

File: .github/workflows/ci-core-reusable.yml [hunk 1]
```diff
@@ -303,6 +304,7 @@ jobs:
       - name: Start services
         run: |
           ci_localnet_up
+          ci_run pre_download_compilers
```
Comment (alexandrst88) [inline .github/workflows/ci-core-reusable.yml:307]: let's make pre_download_compilers with extension .sh? pre_download_compilers.sh

**Review by alexandrst88 — CHANGES_REQUESTED** [review]

File: bin/pre_download_compilers [hunk 2]
```diff
@@ -0,0 +1,44 @@
+#!/usr/bin/env bash
+
```
Comment (alexandrst88) [inline bin/pre_download_compilers:2]: please add set -e

File: bin/pre_download_compilers [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+# This ./cache/hardhat-nodejs is coming from the env-paths module
+# that hardhat is using.
+COMPILER_DIR=/root/.cache/hardhat-nodejs/compilers-v2
+mkdir -p $COMPILER_DIR/linux-amd64
```
Comment (alexandrst88) [inline bin/pre_download_compilers:6]: you could make mkdir -p $COMPILER_DIR/{/linux-amd64,/vyper/linux,/zksolc,/zkvyper}

File: bin/pre_download_compilers [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+wget -nv -O $COMPILER_DIR/zkvyper/compilerVersionInfo.json  "https://raw.githubusercontent.com/matter-labs/zkvyper-bin/main/version.json"
+
+(for ver in v1.3.13; do wget -nv -O $COMPILER_DIR/zkvyper/zkvyper-$ver  https://raw.githubusercontent.com/matter-labs/zkvyper-bin/main/linux-amd64/zkvyper-linux-amd64-musl-$ver; done)
+
+
+# This matches VYPER_RELEASES_MIRROR_URL from hardhat-vyper
+wget -nv -O $COMPILER_DIR/vyper/linux/list.json https://vyper-releases-mirror.hardhat.org/list.json
+
+# Currently we only use 0.3.10 release of vyper compiler (path taken from the list.json above)
+wget -nv -O $COMPILER_DIR/vyper/linux/0.3.10 https://github.com/vyperlang/vyper/releases/download/v0.3.10/vyper.0.3.10%2Bcommit.91361694.linux
```
Comment (alexandrst88) [inline bin/pre_download_compilers:34]: why can't we use https://raw.githubusercontent.com use github cdn host for vyperlang?

File: bin/pre_download_compilers (hunk 2 above)
Comment (tomg10) [reply bin/pre_download_compilers:2]: corrected

File: bin/pre_download_compilers (hunk 3 above)
Comment (tomg10) [reply bin/pre_download_compilers:6]: nice, corrected

File: bin/pre_download_compilers (hunk 4 above)
Comment (tomg10) [reply bin/pre_download_compilers:34]: I wasn't able to make it work for releases, they are not a part of the repo, github redirects the path above to:
`https://objects.githubusercontent.com/github-production-release-asset-2e65be/73461676/c1dd89db-ee80-4888-8d99-f68488ab4331?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=releaseassetproduction%2F20240529%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20240529T220521Z&X-Amz-Expires=300&X-Amz-Signature=a5b1dfb06430ab5707231016f801d12baeb1aae96f3901b1ee84cec963f9871d&X-Amz-SignedHeaders=host&actor_id=18558509&key_id=0&repo_id=73461676&response-content-disposition=attachment%3B%20filename%3Dvyper.0.3.10%2Bcommit.91361694.linux&response-content-type=application%2Foctet-stream`

File: .github/workflows/ci-core-reusable.yml (hunk 1 above)
Comment (tomg10) [reply .github/workflows/ci-core-reusable.yml:307]: corrected

**Review by alexandrst88 — APPROVED** [review]

**Pull request merged**